FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/pkcs11_operations.c

```diff
@@ -0,0 +1,1038 @@
+    CK_BBOOL xTrueObject = CK_TRUE;
+    CK_KEY_TYPE xPrivateKeyType = CKK_EC;
+    CK_OBJECT_CLASS xPrivateKeyClass = CKO_PRIVATE_KEY;
+    CK_OBJECT_HANDLE xObjectHandle = CK_INVALID_HANDLE;
+    mbedtls_ecp_keypair * pxKeyPair = ( mbedtls_ecp_keypair * ) pxMbedPkContext->pk_ctx;
+
+    xResult = C_GetFunctionList( &xFunctionList );
+
+    if( xResult != CKR_OK )
+    {
+        LogError( ( "Could not get a PKCS #11 function pointer." ) );
+    }
+    else
+    {
+        pxDPtr = ( CK_BYTE * ) malloc( EC_D_LENGTH );
+
+        if( pxDPtr == NULL )
+        {
+            xResult = CKR_HOST_MEMORY;
+        }
+    }
+
+    if( xResult == CKR_OK )
+    {
+        xMbedResult = mbedtls_mpi_write_binary( &( pxKeyPair->d ), pxDPtr, EC_D_LENGTH );
+
+        if( xMbedResult != 0 )
+        {
+            LogError( ( "Failed to parse EC private key components." ) );
+            xResult = CKR_ATTRIBUTE_VALUE_INVALID;
+        }
+    }
+
+    if( xResult == CKR_OK )
+    {
+        if( pxKeyPair->grp.id == MBEDTLS_ECP_DP_SECP256R1 )
+        {
+            pxEcParamsPtr = ( CK_BYTE * ) ( "\x06\x08" MBEDTLS_OID_EC_GRP_SECP256R1 );
+        }
+        else
+        {
+            xResult = CKR_CURVE_NOT_SUPPORTED;
+        }
+    }
+
+    if( xResult == CKR_OK )
+    {
+        CK_ATTRIBUTE privateKeyTemplate[] =
+        {
+            { CKA_CLASS,     NULL /* &privateKeyClass*/, sizeof( CK_OBJECT_CLASS )      },
+            { CKA_KEY_TYPE,  NULL /* &privateKeyType*/,  sizeof( CK_KEY_TYPE )          },
+            { CKA_LABEL,     ( void * ) pcLabel,         ( CK_ULONG ) strlen( pcLabel ) },
+            { CKA_TOKEN,     NULL /* &trueObject*/,      sizeof( CK_BBOOL )             },
+            { CKA_SIGN,      NULL /* &trueObject*/,      sizeof( CK_BBOOL )             },
+            { CKA_EC_PARAMS, NULL /* ecParamsPtr*/,      EC_PARAMS_LENGTH               },
+            { CKA_VALUE,     NULL /* DPtr*/,             EC_D_LENGTH                    }
+        };
+
+        /* Aggregate initializers must not use the address of an automatic variable. */
+        privateKeyTemplate[ 0 ].pValue = &xPrivateKeyClass;
+        privateKeyTemplate[ 1 ].pValue = &xPrivateKeyType;
+        privateKeyTemplate[ 3 ].pValue = &xTrueObject;
+        privateKeyTemplate[ 4 ].pValue = &xTrueObject;
+        privateKeyTemplate[ 5 ].pValue = pxEcParamsPtr;
+        privateKeyTemplate[ 6 ].pValue = pxDPtr;
+
+        xResult = xFunctionList->C_CreateObject( xSession,
+                                                 ( CK_ATTRIBUTE_PTR ) &privateKeyTemplate,
+                                                 sizeof( privateKeyTemplate ) / sizeof( CK_ATTRIBUTE ),
+                                                 &xObjectHandle );
+    }
+
+    if( pxDPtr != NULL )
+    {
+        free( pxDPtr );
+    }
+
+    return xResult;
+}
+
+/*-----------------------------------------------------------*/
+
+static CK_RV prvProvisionPrivateRSAKey( CK_SESSION_HANDLE xSession,
```

Comment:
Is this function used?

Reply:
It is not, and many others weren't used either - removed in this commit: https://github.com/FreeRTOS/FreeRTOS/pull/724/commits/6ccdaf924193ba77b8b2efc6df4c10409982534c